### ISSUE TITLE
Create docs for new integration.

### ISSUE THREAD

**Issue by apaperclip** [mww012/home-assistant#6]

**Comment by apaperclip** [comment]
@mww012   I'm working on the docs and I had a question.

I see the friendly_name matches the name, if not null. Should friendly_name be blank by default or always set to something by default so it is consistent?

**Comment by apaperclip** [comment]
Draft integration page docs. Feedback welcome

https://github.com/apaperclip/home-assistant.io/blob/aruba_instant_docs/source/_integrations/aruba_instant.markdown

**Comment by mww012** [comment]
That looks awesome!  The only thing I would add is, when you mention it doesn't work with ArubaOS, I'd also add that it doesn't work with the InstantOn product line either. 

As far as the friendly_name thing, I'm not actually adding that. HA is doing that on its own as the entity is created. I'm not sure what best practices are around that. Do you know what the behavior is in other components?

**Comment by apaperclip** [comment]
Good point thanks. I've added that for clarification.

For friendly_name I checked around my system and it seems just about everything with config flow has it. Its usually a pretty version of the entity name. In this case it may not make sense to use the derived entity name as the friendly name because it is not exactly friendly in comparison to those with names. _device_tracker.aruba_instant_ab_10_76_56_34_12_ vs _device_tracker.chromecast_ .

Another difference from the legacy tracker is the support for picture, icon, and gravatar. I dont know if these are being deprecated in favor of the person entity or what. I'll see what I can find out.

**Comment by mww012** [comment]
Hey, things have been a little chaotic here over the past week. The friendly_name takes the "device_tracker.aruba_instant_ab_10_76_56_34_12" format when the Instant cluster doesn't know the name of the device.  Int that case we pass the name field as None and HA adds that derived name by default.  I'm not opposed to changing the name  assigned a placeholder name in this instant but I don't know what we'd use.  Any thoughts?

Also, I just pushed a new commit to my dev branch.  It adds the Optional config flow.  You should be able to enable/disable entities by selecting the clients you want to track.  Give it a shot and let me know if you hit any issues.  Next step is to automatically initiate that optional config flow when you first set up then component.

**Comment by apaperclip** [comment]
I'm not sure the naming is a huge deal. Backlog wise, its probably pretty low in my opinion.

Thanks for the updates, I'll check em out!

**Comment by apaperclip** [comment]
Looking good!  love the option flow selection of devices.   
Some observations

- I had to remove and re-add the integration otherwise the options flow to select device had a blank drop down. (probably expected)
- The drop down can be cumbersome for a large list. I have 23 devices, so while not a huge list  - maybe a page list would be easier to navigate, especially w/ a type-to-filter.  The devices I'm tracking dont have names I have to search by mac.
- After selecting and submitting the entities were 'unavailable' I think it took a poll for them to update status. Probably expected.
- Should I test  non-restart-based-client-discovery yet?

**Comment by mww012** [comment]
Thanks for the quick feedback!

- Yes, I honestly wasn't sure what the behavior going through the process you did but the blank drop down makes sense.  After re-adding the integration I'd expect everything to behave normally.
- I agree that the dropdown is cumbersome.  the `cv.multi_select()` method, that creates the list, automatically renders like that.  I like the idea of a filter.  I'll see what I can learn about that.  Maybe it's something we can pull off.  Do you know of an integration that uses a filter?  It'd be nice to have an example.
- The `unavailable` state is expected right now.  The coordinator that handles the refresh is in `device_tracker.py` where the listener method called when the options flow is finished is in `__init__.py`.  The end result is that the coordinator isn't directly available to force a refresh when the options flow is complete.  It makes sense to move the coordinator into `__init__.py` because it can be used to refresh more than just the device tracker.  I wanted to get something working and get your take though so I decided to wait on moving that.  Definitely needs to be done though.
- A non-restart based discovery should work just fine.  Give it a shot.

**Comment by apaperclip** [comment]
No problem. Please don't take them as issues or bugs. Just sharing thoughts for consideration and to let you know what i'm seeing.

I will test do the non-restart test tonight.  

When do you think you'll submit your first PR upstream?
Do you have any other thoughts for the docs? I want to get that PR submitted but want to time it somewhat close to the code PR.

**Comment by apaperclip** [comment]
On testing of the non-restart adds.  No luck right now, here's what I did.

To get to a undiscovered/tracked state, 

1. I turned wifi off on my phone and waited for the device to go 'not_home'.
2. I then when into the integration and device screen and tried to delete the entity, but it would not let me. It said "entities can only be removed if the integration is no longer providing the entity, sometimes you have to restart before you can remove"
3. I restarted HA and now the entity was able to be removed.
4. I waited a few minutes then turned wifi back on my phone (confirmed aruba saw it)
5. I waited ~5 minutes, checking the options device drop down and the entity screen and my mac never showed up.
6. Restarted HA and it was now populated in the drop down.

I wonder if there only a single type of 'available' for an integration provided an entity that prevents distinguishing from tracked vs simply added.

**Comment by mww012** [comment]
I'll dig into that and hopefully find the root issue.  I tested and got the same results you did.  I did test it once and it "worked" but it may have just been a coincidence or I misread the results.

As far as when to push the PR, I guess it's as soon as we feel it's ready for prime time.  I'd like to get the non-restart adds working and make the Options config flow automatic as part of the initial flow when adding the component.  After that it's making sure we're compliant with Black/PEP8/etc. I forget exactly what linters they want followed.  Aside from that, do you have anything that you think needs to be done before we can start pushing toward a PR?

**Comment by apaperclip** [comment]
sounds good, good luck.  happy to test more as you iterate.

I think that's good todo list. I know they like PRs being small and manageable, think MVP.

**Comment by mww012** [comment]
Just wanted to give an update.  I've almost got things working.  I've hit one wall I haven't found a way around yet.  When you enable a client via the Options flow, you're not supposed to leverage the 'disabled_by' property to make that happen.  [See here.](https://developers.home-assistant.io/docs/entity_registry_disabled_by#integrations-offering-options-to-control-disabled_by)  I have yet to figure out the proper way to make that work.  I've been looking at the Unifi component but it's not clear to me exactly what they're doing.  Probably my lack of experience with development.  I'll figure it out eventually.

**Comment by apaperclip** [comment]
That's interesting. From that link it reads like the client should be simply deleted from the entity and device registries?  And I guess the next time its discovered the user will have the option to enable it and added it back?

**Comment by mww012** [comment]
That was my thoughts too!  I've been trying to recreate the entity when you enable it. So far I can remove the entity but when I try to added it back via the ```async_add_entity``` method it never fully gets recreated. I know I'm missing something but can't nail down what yet.

**Comment by apaperclip** [comment]
Hey @mww012   - hope all is well, just seeing how things are going well. Anything I can help with?

**Comment by mww012** [comment]
Hey! Sorry I've been a little MIA. Work has gotten busy. I think I'm close to having this worked out. If you want to test I can push my changes up to my repo tomorrow. I can successfully add new entities now but something's getting weird when I try to remove one.

**Comment by apaperclip** [comment]
no need to apologize, I was just reaching out.  Up to you if you want to push, I"m fine to wait if you want more time.

**Comment by mww012** [comment]
Holy crap, I think it works!  Lol, I just pushed the latest round of changes to my dev branch. I've been able to add/remove entities without throwing a list of errors.  I do know there are a few cases that I need to fix.  One issue is that if you try to run the Options flow, when a client that was previously listed is no longer connected to the wireless, it throws an error because it's still looking for information about that client to display to the user.  I'm working on the logic for that next.

Definitely progress though!

**Comment by apaperclip** [comment]
Awesome! I've updated my fork and local git repo. I'll test these 3 scenarios to start.

1. restart HA while phone away 
2. associate new phone while HA is running
3. general entity mgmt (delete, unselect, hide ,etc)

**Comment by mww012** [comment]
Hey, I pushed a new commit up today.  I ran into a problem if HA got rebooted when a tracked entity wasn't connected to the Instant cluster.  It will now properly recreate the entity with the name "Unknown" for now until it successfully gets and update from the cluster containing info for that client.

Next thing I ran into is that we build the list of entities from the last client update from the Instant cluster.  If you have a tracked entity that's no longer connected, it won't show up in the list of trackable clients.  The end result is that you can't untrack that client.  I'm still working with possible options there.  One thing is to have a "discovered clients" list, when you're selecting clients you want to track, and also have "tracked clients" list that gives all currently tracked clients.  We'd build that list from the currently tracked clients so you'll always get a complete list of clients.

Any thoughts?

**Comment by apaperclip** [comment]
Nice!  I ran into that one too and retesting shows it's matches your finding.

1. Associate client
2. Enable tracking for it
3. Disable wifi on client
4. Restart HA
5. The entity_id is correct and state is correct, but the friendly name gets changed to Unknown.  I noticed this because my entity card had the name unkown now. Thet card behavior where it will show the friendly name instead of the entity_id. 
https://www.home-assistant.io/lovelace/entities/#name

I'm not sure what options there ares but it doesn't seem ideal.  What do you think?

Regarding your second observation.  That approach sounds good to me but I do not know what HA best practices are. In the known_devices approach the file would just get longer and longer until you manually edited out a discovered device.  (sounds like it becomes "disabled" in config_flow-ese.

From the other angle, if a device is unassociated it will not appear in the drop down, but at least it stays selected so when hitting submit it doesn't get removed from the tracked entities list. but it does get removed from dev-tools>states.


For my other two test cases

2. Associating a new device was picked up and worked like expected. I had to exit the expanded drop down and go back into 'options' for the integration for the list to show it. Not an issue just sharing results.


3.  Disabling/enabling worked as expected.  Name became unknown which matched other integrations (e.g. wled) after enabling the next poll restored the name.  I was unable to delete 

I was unable to delete a tracked client that was no longer associated. I had to restart HA first.  This _may_ be standard HA behavior. I'm still learning about disable/delete.

If there are any other things you want me to test or are ready to retest just let me know.
Great work again, thank you!

**Comment by apaperclip** [comment]
I just realized something - with hacktoberfest coming up maybe you could earn a tshirt!

**Comment by mww012** [comment]
Lol, that would be nice!  I've got a push that I'm going to push to my repo this morning.  I've made another branch to test the available/tracked idea.  Just pushed it up.  You'll need to update your clone and then checkout the "discover-and-tracked-clients" branch.  There were a pretty hefty chunk of changes to how things work under the hood.  I actually like it this way better.  Give it a shot and let me know what issues you hit.

**Comment by apaperclip** [comment]
The  "discover-and-tracked-clients"  branch is looking good.  I like the two lists, available and tracked.  

A restart while the client is not associated results in it becoming unknown.  Once the client is back online, the next poll switches it back to the name. I don't know what is involved to change this behavior, a restart while a tracked client is associated is likely to 
happen.  I dont think it would break any automations since they will use the entity name and not the friendly name.

When a client is no longer associated, it does not appear in the tracked list.  The only issue I could see this causing is how would you remove a client that's not currently associated.  i.e. get new phone, or only tracking someone temporarily.  As a possible option I tried the stop tracking all clients while a client wasn't associated.  This blew away all the entities as expected, including the one not currently associated. This feels a little heavy handed but would technically work.

I'll do some more testing, I just wanted to get you an initial response.

**Comment by mww012** [comment]
@apaperclip sorry, this comment turned out to be a small novel. 

I did see notice the same issue you ran into regarding a tracked client not showing up in the list if it wasn't associated. I pushed a fix for that a bit earlier today.

As far as the client being named "Unknown", I know what causes it but I'm not sure what the best fix is.  Here's what's going on to the best of my knowledge...

The Coordinator is an HA construct responsible for updating all the information relevant to acomponent. We wrap the Coordinator class into an InstantCoordinator class so we can manipulate it to properly update all the info from the Instant cluster. It keeps a list of all the clients and APs and the associated attributes.  That's where we're pulling the friendly name from, as well as nearly every other attribute. 

The issue comes from the fact that the coordinator only ever exists in memory. So when HA restarts, it dumps the current state of the coordinator and starts over. If we have a tracked client that is no longer connected, after a reboot the coordinator no longer has any information about that client. So when HA goes to create that entity it can't find a friendly name in the coordinators data. That is corrected on the next coordinator refresh after the client has reconnected. 

As best I can tell, the config_entry class is the only part of the components that survives a reboot, which obviously doesn't cover our coordinator info. That's where we're keeping a list of all the mac addresses that we're tracking so that HA knows what entities to recreate when it's rebooted. The probably I ran into is that the data in the config_entry needs to be hashable and while a set() is, a dict() isn't. So it can basically hold a list but not a key/value pair. That dict is what we would need to recreate the entity with all its attributes if the client happened to not be connected when HA was rebooted.

I don't like the fact that it's behaving this way but I haven't found a workable solution. It's definitely something I want to correct but I'm not sure if it's needed now. 

If we're sticking with the idea of MVP for a first PR then does this issue qualify?  If you think it's worth diving into we definitely can but I don't know how deep that rabbit whole will go. 

Thanks for being patient with me on this. It's been a lot of fun to learn and has definitely stretched my Python skill set. Work has started to get a bit busy so I'm not putting as much time into this as I'd like.  We'll get it working though, it's just taking longer than I wanted.

**Comment by apaperclip** [comment]
@mww012 Thanks for the detail reply, that helps a lot.

I see what you mean.  I'm trying to think of other integrations that would have the same issue.  mqtt comes to mind but we use retained to address that.  This may just be how it is so that HA can always represent fresh data. Although there is the new feature to maintain state through a reboot until it get refreshed.

I agree with you about mvp, I think this can be addressed once we get he code into a release.

I completely understand and am thrilled to be where we are. Is that for open issues then? (  I'll test the one you fixed most recently this weekend. )

**Comment by mww012** [comment]
@apaperclip as far as I know it is.  I haven't found any more glaring issues.  If we're ready for a PR to the main HA dev branch I just need to get it to pass the dev tests.  We should be close to passing there.  May need to format some stuff but hopefully that'll be about it.

**Comment by apaperclip** [comment]
Great! I'll review the docs and make any adjustments.   When you are ready we can cross reference all the PRs (logo, docs, code) and submit.

**Comment by mww012** [comment]
@apaperclip just wanted to give you a heads up.  I did something that hosed up my git repo.  No idea exactly what I did so I don't know how to undo it.  I think I'm going to kill my fork of HA and recreated it.  At that point I'll put all my changes back into a new "Aruba Instant" branch because that's what I should have done from the start.  Then hopefully we'll be ready for a PR!

**Comment by mww012** [comment]
Just realized that if I did that it would kill this issue.  I'll just Rebase back to when I forked my branch and that will basically let me start fresh.

**Comment by apaperclip** [comment]
oh no! have you been able to get back up and running?

**Comment by mww012** [comment]
I think I can get the repo back in order.  Now my issue is that when I try to use the dev container in VSCode my CPU spikes to 75% and stays there.  I've tried rebuilding, completely deleting the docker image and re-pulling it, no luck.  Everything that's running inside that container is so slow it's almost unusable.  The weird thing is that the container can be on and running and everything is fine but when I open VSCode and attach to the container, all the utilization goes way up.

**Comment by mww012** [comment]
I think I actually have it all fixed!  As best I can tell, I've been able to pull all of our edits into an aruba_instant branch so we're following HA procedure properly.

**Comment by apaperclip** [comment]
awesome!  glad to hear it.  Is there anything you want me to check?

**Comment by mww012** [comment]
I just made the final edits for testing/linting/etc. and pushed to the aruba_instant branch.  If you can, just run it and make sure I didn't do something dumb.  It still works for me and it was all minor changes like doc string formats, unused imports, etc.

If you're good then I think we're FINALLY ready for a PR.  Took a lot longer than I expected but it's done.  At least phase 1 is done.  When I started on this I didn't expect it to be so involved.  The hass_infinitive integration I wrote was before the UI config was implemented and wasn't nearly as complex.  So much has changed I was basically starting from scratch.

**Comment by apaperclip** [comment]
you got it!

I can't thank you enough for all the time and work you put into this.  This is a huge step forward for the integration.

**Comment by apaperclip** [comment]
Testing is going well, a few of observations

- Weren't there two lists now in the integration options? One that showed tracked clients and the other that showed available? I may have missed an update on that.  This affected me when I went to add new device to be tracked while a tracked device was offline.  Since the offline device wasn't in the list, it was removed from tracking (and as an entity) when I clicked submit.  

- bit of a corner case and not a show stopper I would say  - An entities card with multiple tracked clients will flash them all as unavailable when reenabling a single previously disabled tracked client.

To recreate
1.  Track 2 clients
2.  Disable one in the entities screen
3. Dashboard card should show that entity as unavailable. The other one should be fine
4. Reenable the entity
5. Watch dashboard,  both tracked clients will show the yellow unviable for a second or two in the dashboard then return to normal.

-  I was doing my restart-while-not-connected test to see if the tracked client showed as unavailable in the dashboard until the client rejoined.  That behavior was the same, but I decided to override the name. This behaved better. The entity did not go unavailable in the dashboard, but displayed normally and away through a reboot. Not a new issue or anything we didn't know, except the override name part was a nice surprise.

**Comment by mww012** [comment]
I'm still seeing both lists there.  Did you checkout the aruba_instant branch?  That's where all the new changes should be.

As for the entity card showing tracked entities as unavailable, that happens because when you submit the options config flow with it basically unloads and reloads the component in the background.  That was the only way I could figure out how to get it working properly.  Hopefully there's a better option going forward.  Just haven't found it.

**Comment by apaperclip** [comment]
arg, i flaked on that.  

All looks good, nothing new to report!

**Comment by mww012** [comment]
Nice!  I think we're ready for a PR.  I did look through the docs and I've got a few suggestions but it's basically good to go.

- Somewhere toward the top I'd put that this should work for Instant Version 8.5+.  We should probably point anybody using something lower than that to the current Aruba component via a link.
- I've tested on an AP-325 as well so we can add that model to the list.

**Comment by apaperclip** [comment]
ok go for it!   Thanks for the suggestion, I'll review docs tonight and update.

**Comment by apaperclip** [comment]
Updates made to my repo.  There is a little html on the page that doesn't render on github for the notes. Here's what it looks like when fully rendered.  I added the bold for emphasis. I also made a statement about pre 8.5

![image](https://user-images.githubusercontent.com/67401560/96946813-64542300-14af-11eb-9ef2-8a897d965d7a.png)



Any other feedback is welcome,  Once you do the code PR I will do the PR for the icon and docs.

**Comment by mww012** [comment]
Awesome!  Looks good to me.  I was actually prepping to do the PR later this morning.  

One quick question.  How important is the Integration Quality Scale to you?  We're missing one thing from the silver qualification (the raise PlatformNotReady piece).  Aside from that we may be compliant with the Gold level.  Do you think it's worth adding that in or should we revisit it later?  I'm leaning to handling it later but I'm not very committed to that stance.

**Comment by apaperclip** [comment]
Good question - How much work would it be and how long would it take?  

That said, I'm fine if we wait for the sake of getting the pr submitted.

**Comment by apaperclip** [comment]
tagging @mzac in case hes curious how things are going

**Comment by mww012** [comment]
@apaperclip It shouldn't take much work to add that.  I wanted to try over the weekend but didn't get to it.  If I can't figure it out in the next few days we'll go ahead and submit.  We can always add it later.

**Comment by mww012** [comment]
@apaperclip It's done and pushed.  Was easier than I thought.  It's not 100% perfect because I didn't write instantpy to be compatible with asyncio so I had to basically catch an error thrown when instantpy tries to continue even though it couldn't properly log in.  I need to fix that in the future but this will at least let us check the Integration Scale boxes.

**Comment by mww012** [comment]
@apaperclip @mzac We're submitted!  [RP #42442](https://github.com/home-assistant/core/pull/42442)

@apaperclip When you submit the doc updates, shoot me that PR and I'll update mine with that #.

**Comment by apaperclip** [comment]
yesssss

Docs PR https://github.com/home-assistant/home-assistant.io/pull/15429
Brand PR https://github.com/home-assistant/brands/pull/1792

**Comment by mww012** [comment]
We're getting closer, lol.  I'm hitting an issue with the CI checks.  I've fixed nearly everything.  The Black check in particular is being a pain.  All my tests locally are passing but the CI check isn't.  I don't know if they're using a specific version of Black or what is different.  It says you can reproduce their errors locally with `pre-commit run --all-files` but it all passes when I run it on my box.  If I don't figure it out soon I'll ask for some help in the PR.

**Comment by apaperclip** [comment]
I saw some comments about people needing rebase due to black version changing. not sure if that's still relevant.

**Comment by mww012** [comment]
We just passed all the checks!

That was actually my issue. The precommit command is supposed to upgrade black and pylint and all the other linters it's using. For whatever reason it didn't upgrade the first time I ran it.

Now we just need to make sure the CLA's are signed and the. Hopefully we'll get a reviewer.

**Comment by apaperclip** [comment]
Great to hear. I wonder, do I need to sign the CLA too? I only see it on the code PR

**Comment by mww012** [comment]
Yeah, I think because they've linked you via the doc PR, you'll need to sign one as well.

**Comment by apaperclip** [comment]
ok done.  I guess now we wait - they have a huge backlog so it'll probably be next year. Maybe they'll use some release down-time to work through PR reviews.  or be with their families 🙄

**Comment by apaperclip** [comment]
@mww012  so... what are you thoughts on the feedback so far?

**Comment by mww012** [comment]
I get their stance.  It's better for the userbase for us to just go ahead and replace the existing component.  Means more work for us but it was the direction we were eventually heading anyway.  As soon as our branch gets created I'll start working on integrating the old component.

**Comment by apaperclip** [comment]
Hello,

Just wondering how things have been and if you wanted to pick this up again.

**Comment by mww012** [comment]
Hey!  Things are good.  A little busy with work but that's no reason to complain.  I'm glad you messaged me.  I've been procrastinating getting back to this and you were just the push I needed.  I'll start back on it in the next day or so and I'll keep you up with where we are.

I think our goal now is to bring in all the functionality of the original component so we can completely replace it.  One question I'm not sure about is if we should move completely into lovelace for configuration or try to make it compatible with yaml as well.  Any thoughts?

**Comment by apaperclip** [comment]
good to hear!  Thanks for continuing to think about and be willing to work on this.  From Frenck's comments I took away that we should enhance the existing integration which in my brain literally meant update that existing code.  I guess technically we could just replace it all. I know they like  multiple smaller PRs than fewer larger ones.  That could work in our favor too... 

pr1 could be to replace the existing features, 
pr2 could be to add in the api code and an update to yaml to support it.  
pr3 could be to do config flow, etc.  

 I think its up to you and the existing code.


With the groundwork you've established in the new version do you think it'd be hard to incorporate the ssh methods?  I think the current one just does pexpect.

**Comment by apaperclip** [comment]
Hey, its been a while!  I hope things are well.  I was just checking in to see if you still had any time or interest in this project.

**Comment by mzac** [comment]
Not for me anymore sorry... Moved to Ubiquiti access points for now.

**Comment by mww012** [comment]
@apaperclip I still have interest but this ended up taking a back burner for me. My work life has ramped up so my free time has gone to my family and a few other things. 

I'd like to get back to this but it'll be past the end of the year before I can give this some time.
